File: coding-interview-patterns/scala/src/main/scala/patterns/TwoPointers.scala
package patterns
object TwoPointers {
  def twoSumSorted(a:Array[Int], t:Int): Array[Int] = {
    var l=0; var r=a.length-1
    while (l<r){
      val s=a(l)+a(r)
      if (s==t) return Array(l+1,r+1)
      else if (s<t) l+=1 else r-=1
    }
    Array()
  }
  def threeSum(nums:Array[Int]): List[List[Int]] = {
    scala.util.Sorting.quickSort(nums)
    val n=nums.length; val res=scala.collection.mutable.ArrayBuffer[List[Int]]()
    var i=0
    while (i<n-2){
      var l=i+1; var r=n-1
      while (l<r){
        val s=nums(i)+nums(l)+nums(r)
        if (s==0){ res += List(nums(i),nums(l),nums(r)); l+=1; r-=1
          while (l<r && nums(l)==nums(l-1)) l+=1
          while (l<r && nums(r)==nums(r+1)) r-=1
        } else if (s<0) l+=1 else r-=1
      }
      i+=1; while (i<n-2 && nums(i)==nums(i-1)) i+=1
    }
    res.toList
  }
  def removeDuplicatesSorted(a:Array[Int]): Int = {
    if (a.isEmpty) return 0
    var w=1; var r=1
    while (r<a.length){ if (a(r)!=a(w-1)) { a(w)=a(r); w+=1 }; r+=1 }
    w
  }
}

File: coding-interview-patterns/python/patterns/two_pointers.py
def two_sum_sorted(a, t):
    l,r=0,len(a)-1
    while l<r:
        s=a[l]+a[r]
        if s==t: return [l+1,r+1]
        if s<t: l+=1
        else: r-=1
    return []

def three_sum(nums):
    nums.sort(); n=len(nums); res=[]
    for i in range(n-2):
        if i and nums[i]==nums[i-1]: continue
        l,r=i+1,n-1
        while l<r:
            s=nums[i]+nums[l]+nums[r]
            if s==0:
                res.append([nums[i],nums[l],nums[r]]); l+=1; r-=1
                while l<r and nums[l]==nums[l-1]: l+=1
                while l<r and nums[r]==nums[r+1]: r-=1
            elif s<0: l+=1
            else: r-=1
    return res

def remove_duplicates_sorted(a):
    if not a: return 0
    w=1
    for r in range(1,len(a)):
        if a[r]!=a[w-1]: a[w]=a[r]; w+=1
    return w
